### PR TITLE
Add min, max, constrain function for Matrix

### PR DESCRIPTION
**Pull request by kamilritz**
This PR adds the functionality to compute a constrained Matrix with a scalar or another Matrix of the same size. It provides the functionality of min, max and constrain.

@jkflying

### CODE REVIEW

File: matrix/Matrix.hpp [hunk 1]
```diff
@@ -580,6 +580,126 @@ bool isEqual(const Matrix<Type, M, N> &x,
...
+                    m(i,j) = x_lower_bound(i,j);
+                } else if(x(i,j) > x_upper_bound(i,j)) {
+                    m(i,j) = x_upper_bound(i,j);
+                } else {
+                    m(i,j) = x(i,j);
+                }
+            }
+        }
+    }
+    if(should_be_nan) {
```
Comment (jkflying) [inline matrix/Matrix.hpp]: Shouldn't we only set NaN to the values where this is true?

**Comment by jkflying** [comment]
Something I just realised, the semantics of min and max need to match the standard library definition when one of the values is a NaN. This should be tested, so that eg. `min(NAN, 1) == min(Vector2f(NaN,5),1.f)(0)`, and vice versa with argument ordering.

**Comment by kamilritz** [comment]
> Something I just realised, the semantics of min and max need to match the standard library definition when one of the values is a NaN. This should be tested, so that eg. min(NAN, 1) == min(Vector2f(NaN,5),1.f)(0), and vice versa with argument ordering.

@jkflying which implementation would you suggest for the reference min(NAN,1) call?

**Comment by jkflying** [comment]
Preferably `fmin` and `fmax`, but `std::min`/`max` should have the same semantics.

File: matrix/Matrix.hpp [hunk 2]
```diff
@@ -580,6 +580,128 @@ bool isEqual(const Matrix<Type, M, N> &x,
...
+    return (x < y) ? x : y;
+}
+template<typename Type>
+Type max(const Type x, const Type y) {
+    bool x_is_nan = isnan(x);
+    bool y_is_nan = isnan(y);
+    // z > nan for z != nan is required by C the standard
+    if(x_is_nan || y_is_nan) {
+        if(x_is_nan && !y_is_nan) return y;
+        if(!x_is_nan && y_is_nan) return x;
```
Comment (MaEtUgR) [inline matrix/Matrix.hpp]: The second condition does the same as default hence it could be left away e.g. with a comment.

Comment (kamilritz) [reply matrix/Matrix.hpp]: Good observation

Comment (MaEtUgR) [reply matrix/Matrix.hpp]: https://github.com/PX4/Matrix/pull/126

**Pull request merged**